pkg/cmd/cli/cmd/logs.go

```diff
@@ -0,0 +1,179 @@
+  $ %[1]s logs -f 123456-7890 -c ruby-container`
+)
+
+type OpenShiftLogsOptions struct {
+	KubeClient   *kclient.Client
+	OriginClient *client.Client
+
+	Namespace      string
+	ResourceString string
+	ContainerName  string
+	Follow         bool
+	Interactive    bool
+	Previous       bool
+	Out            io.Writer
+}
+
+// NewCmdLogs creates a new pod log command
+func NewCmdLogs(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	o := &OpenShiftLogsOptions{
+		Out:         out,
+		Interactive: true,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "logs [-f] RESOURCE",
+		Short:   "Print the logs for a container in a pod.",
+		Long:    logsLong,
+		Example: fmt.Sprintf(logsExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(o.Complete(f, out, cmd, args))
+
+			cmdutil.CheckErr(o.Validate(cmd))
+
+			cmdutil.CheckErr(o.RunLog())
+		},
+		Aliases: []string{"log"},
+	}
+	cmd.Flags().BoolVarP(&o.Follow, "follow", "f", o.Follow, "Specify if the logs should be streamed.")
+	cmd.Flags().BoolVar(&o.Interactive, "interactive", o.Interactive, "If true, prompt the user for input when required. Default true.")
+	cmd.Flags().BoolVarP(&o.Previous, "previous", "p", o.Previous, "If true, print the logs for the previous instance of the container in a pod if it exists.")
+	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", o.ContainerName, "Container name")
+	return cmd
+}
+
+func (o *OpenShiftLogsOptions) Complete(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args []string) error {
+	switch len(args) {
+	case 0:
+		return cmdutil.UsageError(cmd, "RESOURCE is required for log")
+
+	case 1:
+		o.ResourceString = args[0]
+	case 2:
+		o.ResourceString = args[0]
+		o.ContainerName = args[1]
+
+	default:
+		return cmdutil.UsageError(cmd, "log RESOURCE [CONTAINER]")
+	}
+
+	var err error
+	o.Namespace, _, err = f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	o.OriginClient, o.KubeClient, err = f.Clients()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *OpenShiftLogsOptions) Validate(cmd *cobra.Command) error {
+	return nil
+}
+
+func (o *OpenShiftLogsOptions) RunLog() error {
+	kLogsOptions := &kcmd.LogsOptions{
+		Client: o.KubeClient,
+
+		PodNamespace:  o.Namespace,
+		ContainerName: o.ContainerName,
+		Follow:        o.Follow,
+		Interactive:   o.Interactive,
+		Previous:      o.Previous,
+		Out:           o.Out,
+	}
+
+	resourceType := "pod"
+	resourceName := o.ResourceString
+	tokens := strings.SplitN(o.ResourceString, "/", 2)
+	if len(tokens) == 2 {
+		resourceType = tokens[0]
+		resourceName = tokens[1]
+	}
+	resourceType = strings.ToLower(resourceType)
+
+	// if we're requesting a pod, delegate directly to kubectl logs
+	if (tokens[0] == "pods") || (tokens[0] == "pod") || len(o.ContainerName) > 0 {
+		kLogsOptions.PodName = resourceName
+		return kLogsOptions.RunLog()
+	}
+
+	switch resourceType {
```

Comment:
> This has to be abstracted out into a method on Factory for this to be viable.

By "this", you mean the idea of handling logs for different kinds of resources?  If we could agree on all flags, I definitely agree.  However, things like `--nowait` seem useful and also unlikely to move upstream.  I could also envision flags "this many builds/deployments before current" and "show me logs for a pod in the RC of a DC's current deployment".  I want my `oc logs` command to do all those things.